BZ1962490: Machine-config-controller not generating new machineconfigs for new Kubeletconfig CR

mburke5678 opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1962490

lyman9966 commented on modules/create-a-containerruntimeconfig-crd.adoc:
```diff
@@ -20,15 +20,15 @@ You can modify the following settings by using a `ContainerRuntimeConfig` CR:
 monitors communications between a container manager (such as Podman or CRI-O) and the OCI runtime (such as runc or crun) for a single container.
 
-You can create multiple `ContainerRuntimeConfig` CRs, as needed, with a limit of ten. For the first `ContainerRuntimeConfig` CR, the MCO creates a machine config appended with `containerruntime`. With each subsequent CR, the controller creates a new `containerruntime` machine config with a numeric suffix. For example, if you have a `containerruntime` machine config with a `-2` suffix, the next `c ...
+You should have one `ContainerRuntimeConfig` CR for each machine config pool with all the config changes you want for that pool. You should edit an existing `ContainerRuntimeConfig` CR to modify existing settings or add new settings instead of creating a new CR for each change. It is recommended to create a new CR only to modify a different machine config pool, or for changes that are intended to ...
```
It's better to add this line after the first sentence:
"even if the same content is being applied to all the pools, then you only need one ContainerRuntimeConfig CR for all the pools"

lyman9966 commented on modules/create-a-containerruntimeconfig-crd.adoc:
```diff
@@ -20,15 +20,15 @@ You can modify the following settings by using a `ContainerRuntimeConfig` CR:
+You should have one `ContainerRuntimeConfig` CR for each machine config pool with all the config changes you want for that pool. You should edit an existing `ContainerRuntimeConfig` CR to modify existing settings or add new settings instead of creating a new CR for each change. It is recommended to create a new CR only to modify a different machine config pool, or for changes that are intended to ...
 
-You can also edit an existing `ContainerRuntimeConfig` CRs to edit existing settings or add new settings instead of creating a new CR. It is recommended to create a new CR to modify a different machine config pool, rather than add additional labels to an existing CR.
+You can create multiple `ContainerRuntimeConfig` CRs, as needed, with a limit of ten. For the first `ContainerRuntimeConfig` CR, the MCO creates a machine config appended with `containerruntime`. With each subsequent CR, the controller creates a new `containerruntime` machine config with a numeric suffix. For example, if you have a `containerruntime` machine config with a `-2` suffix, the next `c ...
```
you'd better clarify that it is not 10 per pool, it is 10 overall. Customers may ignore this.

lyman9966 commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,7 +8,18 @@
-`kubelet-config-controller` added to the Machine Config Controller (MCC). This allows you to create a `KubeletConfig` custom resource (CR) to edit the kubelet parameters.
+`kubelet-config-controller` added to the Machine Config Controller (MCC). This allows you to use a `KubeletConfig` custom resource (CR) to edit the kubelet parameters.
+
+You should have one `ContainerRuntimeConfig` CR for each machine config pool with all the config changes you want for that pool. You should edit an existing `ContainerRuntimeConfig` CR to modify existing settings or add new settings instead of creating a new CR for each change. It is recommended to create a new CR only to modify a different machine config pool, or for changes that are intended to ...
```
please replace all "ContainerRuntimeConfig" with "KubeletConfig"

lyman9966 commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,7 +8,18 @@
+
+You should have one `ContainerRuntimeConfig` CR for each machine config pool with all the config changes you want for that pool. You should edit an existing `ContainerRuntimeConfig` CR to modify existing settings or add new settings instead of creating a new CR for each change. It is recommended to create a new CR only to modify a different machine config pool, or for changes that are intended to ...
+
+You can create multiple `KubeletConfig` CRs, as needed, with a limit of ten. For the first `KubeletConfig` CR, the MCO creates a machine config appended with `kubelet`. With each subsequent CR, the controller creates a new `kubelet` machine config with a numeric suffix. For example, if you have a `kubelet` machine config with a -2 suffix, the next `kubelet` machine config is appended with -3.
```
Also you'd better clarify that it is not 10 per pool, it is 10 overall. Customers may ignore this.

lyman9966 commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,7 +8,18 @@
+
+[NOTE]
+====
+If you have a machine config with a `containerruntime-9` suffix, and you create another `ContainerRuntimeConfig` CR, a new machine config is not created, even if there are fewer than 10 `containerruntime` machine configs.
```
please  replace "ContainerRuntimeConfig" with "KubeletConfig", and replace "containerruntime" with "kubelet"

mburke5678 commented:
@lyman9966 Thank you for your comments!

lyman9966 commented:
@mburke5678  I'm afriaid there are some mistakes in your modification. FYI : https://deploy-preview-33056--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/machine-configuration-tasks.html#create-a-containerruntimeconfig_post-install-machine-configuration-tasks
We create a ContainerRuntimeConfig to edit CRI-O parameters, yet in this chapter you write KubeletConfig everywhere.

mburke5678 commented:
@lyman9966 I apologize for wasting your time. I made a mistake when cutting-and-pasting "ContainerRuntimeConfig" and "KubeletConfig" [as you suggested for the other file](https://github.com/openshift/openshift-docs/pull/33056#discussion_r652424242) in this PR. I have reversed these changes. 

Also, I want to add a list of machine configs with a `99-worker-generated-kubelet` machine config. The first kubeletconfig I created resulted in `99-worker-generated-kubelet-1`. Is this correct? The first machine config for a ContainerRuntimeConfig does not have a number: `99-worker-generated-containerruntime`.

See new changes: https://github.com/openshift/openshift-docs/pull/33056/commits/ddf7333e5c4b82740c57420b63081bd22159a3d9#diff-f8526981ec89a35c745acc038fb1fa91f5090e9d2f36bc9ac144c4452b67161aR53

lyman9966 commented:
@mburke5678 the first kubeletconfig should be `99-worker-generated-kubelet`

lyman9966 commented:
@mburke5678 I comment here: https://github.com/openshift/openshift-docs/commit/ddf7333e5c4b82740c57420b63081bd22159a3d9#

mburke5678 commented:
@lyman9966 I made the changes as suggested. I added a prerequisite to look for a label in the MCP and add one if not present, which I copied from another module in this topis. And added some more details that I saw as I tested the procedure. PTAL and thank you for all your help here.

lyman9966 commented:
/lgtm, thanks @mburke5678

adellape commented on modules/create-a-containerruntimeconfig-crd.adoc:
```diff
@@ -3,7 +3,7 @@
 
 [id="create-a-containerruntimeconfig_{context}"]
-= Creating a ContainerRuntimeConfig CR to edit CRI-O parameters
+= Creating a `ContainerRuntimeConfig` CR to edit CRI-O parameters
```
Guidelines say no backticks in assembly/module headings (but are OK in block titles :shrug:).

adellape commented on modules/create-a-containerruntimeconfig-crd.adoc:
```diff
@@ -20,15 +20,17 @@ You can modify the following settings by using a `ContainerRuntimeConfig` CR:
-You can also edit an existing `ContainerRuntimeConfig` CRs to edit existing settings or add new settings instead of creating a new CR. It is recommended to create a new CR to modify a different machine config pool, rather than add additional labels to an existing CR.
+You should edit an existing `ContainerRuntimeConfig` CR to modify existing settings or add new settings instead of creating a new CR for each change. It is recommended to create a new `ContainerRuntimeConfig` CR only to modify a different machine config pool, or for changes that are intended to be temporary so that you can revert the changes.
+
+You can create multiple `ContainerRuntimeConfig` CRs, as needed, with a limit of ten per cluster. For the first `ContainerRuntimeConfig` CR, the MCO creates a machine config appended with `containerruntime`. With each subsequent CR, the controller creates a new `containerruntime` machine config with a numeric suffix. For example, if you have a `containerruntime` machine config with a `-2` suffix, ...
```
s/ten/10/

IBM SG says spelling out is only an option for under 10.

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,35 +8,114 @@
+
+You should edit an existing `KubeletConfig` CR to modify existing settings or add new settings instead of creating a new CR for each change. It is recommended to create a new CR only to modify a different machine config pool, or for changes that are intended to be temporary so that you can revert the changes.
+
+You can create multiple `KubeletConfig` CRs, as needed, with a limit of ten per cluster. For the first `KubeletConfig` CR, the MCO creates a machine config appended with `kubelet`. With each subsequent CR, the controller creates a new `kubelet` machine config with a numeric suffix. For example, if you have a `kubelet` machine config with a `-2` suffix, the next `kubelet` machine config is appende ...
```
s/ten/10/

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,35 +8,114 @@
+If you have a machine config with a `kubelet-9` suffix, and you create another `KubeletConfig` CR, a new machine config is not created, even if there are fewer than 10 `kubelet` machine configs.
+====
+
+.Example showing a `KubeletConfig` CR
```
Probably could lose "showing a" in the title (here and in the next instance below).

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,35 +8,114 @@
+
+.Prerequisites
+
+. Obtain the label associated with the static `MachineConfigPool` CR for the type of node you want to configure.
```
This is sort of a single-step procedure, though it has two options as substeps. Maybe make them all unordered `*` and `**` instead of ordered `.` and `..`?

mburke5678 replied:
I copied this from another module in the assembly. Maybe I can change both.

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,35 +8,114 @@
 .Procedure
 
-. Run:
+The following procedure is an example to show how to configure the maximum number of pods per node on the worker nodes.
```
I think this could go up before the `.Prerequisites`.

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,35 +8,114 @@
 [source,terminal]
 ----
-# oc describe node <node-ip> | grep Allocatable -A6
+# oc describe node <node_name>
```
Was the switch from the earlier `$` prompt to `#` here (and the next instance) intentional?

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -8,35 +8,114 @@
+# oc describe node ci-ln-5grqprb-f76d1-ncnqq-worker-a-mdv94
 ----
 +
+Look for `value: pods: <value>` in the `Allocatable` stanza.
```
s/stanza./stanza:/

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -49,7 +128,7 @@ Allocatable:
 ----
 
-. To set the max pods per node on the worker nodes, create a custom resource file that contains the kubelet configuration. For example, `change-maxPods-cr.yaml`:
+. Set the maximum pods per node on the worker nodes by creating a custom resource file that contains the kubelet configuration. For example, `change-maxPods-cr.yaml`:
```
Suggest moving the "For example" fragment to just be the block title:

```
that contains the kubelet configuration:
+
.Example `change-maxPods-cr.yaml` file
[source,yaml]
----
```

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -60,12 +139,17 @@ metadata:
-    maxPods: 500
+    maxPods: 500 <2>
 ----
+<1> Enter the label from the Machine Config Pool.
```
Looks like this is the only init cap instance, so:

s/Machine Config Pool/machine config pool/

adellape commented on modules/create-a-kubeletconfig-crd-to-edit-kubelet-parameters.adoc:
```diff
@@ -60,12 +139,17 @@ metadata:
-The rate at which the kubelet talks to the API server depends on queries per second (QPS) and burst values. The default values, `50` for `kubeAPIQPS` and `100` for `kubeAPIBurst`, are good enough if there are limited pods running on each node. Updating the kubelet QPS and burst rates is recommended if there are enough CPU and memory resources on the node:
+[NOTE]
+====
+The rate at which the kubelet talks to the API server depends on queries per second (QPS) and burst values. The default values, `50` for `kubeAPIQPS` and `100` for `kubeAPIBurst`, are sufficient if there are limited pods running on each node. It is recommended to update the kubelet QPS and burst rates if there are enough CPU and memory resources on the node:
```
This NOTE ends with a colon ("the node:") but there's nothing after it within the admonition box. Is the code block that follows it supposed to be up in the NOTE, or should that just be "the node."?

mburke5678 commented:
/cherrypick enterprise-4.7
